server/leafnode.go

```diff
@@ -0,0 +1,1074 @@
+			return
+		}
+		defer wipeSlice(contents)
+		items := credsRe.FindAllSubmatch(contents, -1)
+		if len(items) < 2 {
+			c.Errorf("Credentials file malformed")
+			return
+		}
+		// First result should be the user JWT.
+		// We copy here so that the file containing the seed will be wiped appropriately.
+		raw := items[0][1]
+		tmp := make([]byte, len(raw))
+		copy(tmp, raw)
+		// Seed is second item.
+		kp, err := nkeys.FromSeed(items[1][1])
+		if err != nil {
+			c.Errorf("Credentials file has malformed seed")
+			return
+		}
+		// Wipe our key on exit.
+		defer kp.Wipe()
+
+		sigraw, _ := kp.Sign(c.nonce)
+		sig := base64.RawURLEncoding.EncodeToString(sigraw)
+		cinfo.JWT = string(tmp)
+		cinfo.Sig = sig
+	} else if userInfo := c.leaf.remote.URL.User; userInfo != nil {
+		cinfo.User = userInfo.Username()
+		pass, _ := userInfo.Password()
+		cinfo.Pass = pass
+	}
+
+	b, err := json.Marshal(cinfo)
+	if err != nil {
+		c.Errorf("Error marshaling CONNECT to route: %v\n", err)
+		c.closeConnection(ProtocolViolation)
+		return
+	}
+	c.sendProto([]byte(fmt.Sprintf(ConProto, b)), true)
+}
+
+// Called when an inbound leafnode connection is either accepted.
+func (s *Server) createLeafNode(conn net.Conn, remote *RemoteLeafOpts) *client {
+	// Snapshot server options.
+	opts := s.getOpts()
+
+	maxPay := int32(opts.MaxPayload)
+	maxSubs := int32(opts.MaxSubs)
+	// For system, maxSubs of 0 means unlimited, so re-adjust here.
+	if maxSubs == 0 {
+		maxSubs = -1
+	}
+	now := time.Now()
+
+	c := &client{srv: s, nc: conn, kind: LEAF, opts: defaultOpts, mpay: maxPay, msubs: maxSubs, start: now, last: now}
+	c.leaf = &leaf{smap: map[string]int32{}}
+
+	// Determines if we are soliciting the connection or not.
+	var solicited bool
+
+	if remote != nil {
+		solicited = true
+		// Users can bind to any local account, if its empty
+		// we will assume the $G account.
+		if remote.LocalAccount == "" {
+			remote.LocalAccount = globalAccountName
+		}
+		// FIXME(dlc) - Make this resolve at startup.
+		c.acc, _ = s.LookupAccount(remote.LocalAccount)
+		// Make sure we register with the account here.
+		c.registerWithAccount(c.acc)
+		c.leaf.remote = remote
+	}
+
+	// Grab server variables
+	s.mu.Lock()
+	info := s.leafNodeInfo
+	s.mu.Unlock()
+
+	// Grab lock
+	c.mu.Lock()
+
+	c.initClient()
+
+	c.Debugf("LeafNode connection created")
+
+	if solicited {
+		// We need to wait here for the info, but not for too long.
+		b := make([]byte, MAX_CONTROL_LINE_SIZE)
+		c.nc.SetReadDeadline(time.Now().Add(time.Second))
+		n, err := c.nc.Read(b)
+		if err != nil {
+			c.mu.Unlock()
+			if err == io.EOF {
+				c.closeConnection(ClientClosed)
+			} else {
+				c.closeConnection(ReadError)
+			}
+			return nil
+		}
+		c.nc.SetReadDeadline(time.Time{})
+
+		c.mu.Unlock()
+		// Error will be handled below, so ignore here.
+		c.parse(b[:n])
+		c.mu.Lock()
+
+		if !c.flags.isSet(infoReceived) {
+			c.mu.Unlock()
+			c.Debugf("Did not get the remote leafnode's INFO, timed-out")
```

Comment:
Would we not had timed out in the `c.nc.Read()` above instead? Instead, I think we could be here just because we receive a partial proto.. c.parse() would error (or save partial) and then we would get here since infoReceived flag is not set and say that we failed due to timeout.. which is not the case.

Reply:
agree